### ISSUE TITLE
Interface documentation

### ISSUE THREAD

**Issue by denizyuret** [FluxML/Zygote.jl#12]
I am trying to understand the interface you are planning. Is there any documentation other than the examples in README? The basic questions for the interfaces you introduce (gradient, derivative, forward):
* How do I mark variables I want gradients wrt? 
* How do I get those gradients out?
* How do I run code with/without gradient calculation?

**Comment by MikeInnes** [comment]
No other docs yet, but I can try to clarify these and figure out what needs documenting explicitly.

In general we specify the args we want using closures, e.g.
```julia
f(a, b) = a*b
gradient(f, 2, 3) # (3, 2)
gradient(x -> f(2, x), 3) # (2,)
```

This is one of those things that makes total sense but is a bit non-obvious at first, so it'd be worth having explicit docs on this.

I'm a bit confused by the other two questions; the gradients are right there, and running without gradients is just `f(2, 3)` as usual. Maybe you're referring to the `Params` API which is a bit different.

It might be useful to see how I'd simulate Knet/AutoGrad's original API with Zygote:

```julia
grad(f, arg = 1) = (x...) -> gradient(f, x...)[arg]

loss(θ, x) = sum(θ[:W] * x .+ θ[:b])

∇loss = grad(loss)

θ = Dict(:W => randn(5,5), :b => randn(5))
# θ = (W = randn(5,5), b = randn(5)) # NamedTuples work too

∇loss(θ, rand(5)) # Dict(:W => ..., :b => ...)
```

**Comment by denizyuret** [comment]
On Fri, Sep 7, 2018 at 10:05 AM Mike J Innes <notifications@github.com>
wrote:

> No other docs yet, but I can try to clarify these and figure out what
> needs documenting explicitly.
>
I think a regular function doc giving the input/output specification would
be sufficient. There are many functions in many packages that serve similar
roles (grad, gradient, derivative etc.) each with a slight difference in
interface.

> In general we specify the args we want using closures, e.g.
>
> f(a, b) = a*bgradient(f, 2, 3) # (3, 2)gradient(x -> f(2, x), 3) # (2,)
>
> This is one of those things that makes total sense but is a bit
> non-obvious at first, so it'd be worth having explicit docs on this.
>
So you get a tuple of gradients for all the args supplied to gradient. Are
there any type restrictions on the args?  Can the arguments be structs?
Does it reconstruct dicts, AnyArrays etc?  Are there any efficiency hacks
(special value for zero grads, dropping irrelevant dictionary keys etc.)
How about the `derivative` interface? etc.


> I'm a bit confused by the other two questions; the gradients are right
> there, and running without gradients is just f(2, 3) as usual. Maybe
> you're referring to the Params API which is a bit different.
>
These are just the standard questions any AD user is going to ask. Yes,
with the gradient interface I agree the answers are clear (except see
questions above).  The low level Params / forward interface may need some
explanation.

It might be useful to see how I'd simulate Knet/AutoGrad's original API
> with Zygote:
>
> grad(f, arg = 1) = (x...) -> gradient(f, x...)[arg]
> loss(θ, x) = sum(θ[:W] * x .+ θ[:b])
>
> ∇loss = grad(loss)
>
> θ = Dict(:W => randn(5,5), :b => randn(5))# θ = (W = randn(5,5), b = randn(5)) # NamedTuples work too
> ∇loss(θ, rand(5)) # Dict(:W => ..., :b => ...)
>
>
I am wondering if we can write a 5 line mini-Zygote example in Julia from
scratch (analogous to Alan's intro AD example), but we can discuss this
offline...

**Comment by MikeInnes** [comment]
Just added [fairly comprehensive](http://fluxml.ai/Zygote.jl/dev/) interface documentation to the docs. Let me know if there's anything unclear in that.
